### PR TITLE
Use most common responses

### PR DESCRIPTION
**Pull request by joshuef**
Dont rely on last response only.
Adds a connection hook manager for easier testing
adds tests

### CODE REVIEW

File: safe_core/src/connection_manager/connection_group.rs [hunk 1]
```diff
@@ -270,16 +272,18 @@ impl Joining {
...
-    hooks: HashMap<MessageId, (Sender<Response>, usize)>, // to be replaced with Accumulator for multiple vaults.
+    hook_manager: ConnectionHookManager,
 }
 
 impl Connected {
     fn new(old_state: Joining) -> Self {
         // trigger the connection future
         let _ = old_state.connection_hook.send(Ok(()));
 
+        let response_threshold = old_state.connected_elders.len() / 2;
```
Comment (lionel1704) [inline safe_core/src/connection_manager/connection_group.rs:283]: Will this be 3 or 4 for 7 elders?

File: safe_core/src/connection_manager/connection_hook_manager.rs [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+
+use futures::sync::oneshot::Sender;
+use log::trace;
+
+use safe_nd::{MessageId, Response};
+
+use std::collections::HashMap;
+
+type ResponseRequiredCount = usize;
+type VoteCount = u8;
```
Comment (lionel1704) [inline safe_core/src/connection_manager/connection_hook_manager.rs:17]: `usize` is usually used for anything `count`-y

File: safe_core/src/connection_manager/connection_hook_manager.rs [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+type VoteMap = HashMap<Response, VoteCount>;
+
+pub struct ConnectionHookManager {
+    /// MessageId to send_future channel map
+    hooks: HashMap<MessageId, (Sender<Response>, VoteMap, ResponseRequiredCount)>,
+    /// Number of responses to aggregate before returning to a client
+    response_threshold: u8,
```
Comment (lionel1704) [inline safe_core/src/connection_manager/connection_hook_manager.rs:24]: Same comment about `usize`

File: safe_core/src/connection_manager/connection_group.rs (hunk 1 above)
Comment (joshuef) [reply safe_core/src/connection_manager/connection_group.rs:283]: It will be however many elders there are. Over two and plus one to ensure majority.

Comment (joshuef) [reply safe_core/src/connection_manager/connection_group.rs:283]: Hmm, yeh this needs that plus one to ensure it. Let me check where I've added that elsewhere, I think that can be clearer 👍

File: safe_core/src/connection_manager/pending_request_manager.rs [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+type VoteMap = HashMap<Response, VoteCount>;
+
+pub struct PendingRequestManager {
+    /// MessageId to send_future channel map
+    requests: HashMap<MessageId, (Sender<Response>, VoteMap, ResponseRequiredCount)>,
+    /// Number of responses to aggregate before returning to a client
+    response_threshold: usize,
+}
+
+/// Manage request requests and their responses
```
Comment (Yoga07) [inline safe_core/src/connection_manager/pending_request_manager.rs:27]: This comment should be corrected.

File: safe_core/src/connection_manager/pending_request_manager.rs [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+                    let _ = vote_map.insert(vote_response, 1);
+                }
+
+                trace!("vote map looks like: {:?}", &vote_map);
+
+                // if 50+% successfull responses, we roll with it.
+                if current_count <= self.response_threshold {
+                    let mut vote_met_threshold = false;
+
+                    let _the_response = &response;
```
Comment (Yoga07) [inline safe_core/src/connection_manager/pending_request_manager.rs:83]: Can be removed

File: safe_core/src/connection_manager/pending_request_manager.rs [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+
+                    for (_response_key, votes) in vote_map.iter() {
+                        if votes >= &self.response_threshold {
+                            // the_response = response_key;
```
Comment (Yoga07) [inline safe_core/src/connection_manager/pending_request_manager.rs:87]: Can be removed

**Pull request merged**